Add examples on how to embed a Flutter screen into an existing SwiftUI app (#7828)
flutter/website: src/development/add-to-app/ios/add-flutter-screen.md

### Create a FlutterEngine

The proper place to create a `FlutterEngine` is specific

> to your host app. 

Author: I think we want to say what we're doing though - I think its important that we mention we're exposing FlutterEngine as an app delegate when the app starts up. Is there a better way to word it that includes this information?

> to your host app. 

Author: Actually never mind I see you add this below!

> to your host app. 

Author: okay!

{% samplecode engine %}

{% sample SwiftUI %}

> As an example, we demonstrate creating a
> `FlutterEngine` inside an `ObservableObject`, 
> and passing that into a `View` using the 
>  `environmentObject()` modifier. 

Reviewer: **issue:** This section should indicate what's done where. This throws in a lot of information with little context.
```suggestion
In this example, we create a `FlutterEngine` object inside a SwiftUI `ObservableObject`. 
We then pass this `FlutterEngine` into a `ContentView` using the 
 `environmentObject()` property. 
```

**suggestion:** It would not hurt to link out to the [SwiftUI docs](https://developer.apple.com/documentation/swiftui/environmentobject) in this case to make sure the dev can follow along.

 **In `MyApp.swift`:**
 <!--code-excerpt "MyApp.swift" title-->
 ```swift
import SwiftUI

> import Flutter
> // Used to connect plugins (only if you have plugins with iOS platform code).
> import FlutterPluginRegistrant

Reviewer: **question:** Which line does this modify? The previous or the following? It reads like it's the previous. If it is the `FlutterPluginRegistrant`, then move this comment to after that line.

**suggestion:** Consider the context. _What_ is used to connect plugins? Also, why use the parenthetical? It can be combined.

```suggestion
import Flutter
import FlutterPluginRegistrant
// The previous library connects plugins with iOS platform code to this app.
```

Author: Hmm I'm more familiar with comments that go before the line that it is describing, but I can make it explicit that we are referring to FlutterPluginRegistrant

class FlutterDependencies: ObservableObject {

>   let flutterEngine = FlutterEngine(name: "my flutter engine")

Reviewer: **suggestion:** Don't label the variable with the same exact label as the Object. That makes it hard to follow. 

```suggestion
  let myFlutterEngine = FlutterEngine(name: "my flutter engine")
```

Author: We use flutterEngine as the variable name in several examples that were already in this document and are available as demo apps in multiple places on GitHub, so I think it makes sense to leave it as is. Are you okay with that or would you still like me to change it?

Reviewer: **reply:** Leave it. If you could create a new issue to address it later, that would be appreciated!

Author: Done!

  init(){

>     // Prepare a Flutter engine in advance.

Reviewer: **question:** In advance of what?

    flutterEngine.run()

>     // Used to connect plugins (only if you have plugins with iOS platform code).

Reviewer: **repeat:** See [line 57](https://github.com/flutter/website/pull/7828/files#diff-16e6938e4d60842765926435d41283877a42dc43768665948e328d2b563739d2R57).

{% samplecode engine %}

{% sample SwiftUI %}
As an example, we demonstrate creating a
`FlutterEngine` inside an `ObservableObject`, 
and passing that into a `View` using the 
 `environmentObject()` modifier. 

 **In `MyApp.swift`:**
 <!--code-excerpt "MyApp.swift" title-->
 ```swift
import SwiftUI
import Flutter
// Used to connect plugins (only if you have plugins with iOS platform code).
import FlutterPluginRegistrant

class FlutterDependencies: ObservableObject {

>   let npsFlutterEngine = FlutterEngine(name: "my flutter engine")

Reviewer 2: nit: maybe just `flutterEngine`? readers may not have the context of newsfeed app (iirc this nps is from the news feed demo app?)

Author: Oops yeah my bad, thanks!

  init(){
    // Prepare a Flutter engine in advance.
    npsFlutterEngine.run()
    // Used to connect plugins (only if you have plugins with iOS platform code).
    GeneratedPluginRegistrant.register(with: self.npsFlutterEngine);
  }
}

@main
struct MyApp: App {
  // flutterDependencies will be injected using EnvironmentObject.
  @StateObject var flutterDependencies = FlutterDependencies()

>     var body: some Scene {

Reviewer 2: indentation

  }
}
```
{% sample UIKit-ObjC %}
As an example, we demonstrate creating a
`FlutterEngine`, exposed as a property, on app startup in

> the app delegate.

Reviewer: **repeat:** See [line 49](https://github.com/flutter/website/pull/7828/files#diff-16e6938e4d60842765926435d41283877a42dc43768665948e328d2b563739d2R49).

### Show a FlutterViewController with your FlutterEngine

{% samplecode vc %}

{% sample SwiftUI %}
The following example shows a generic `ContentView` with a
`Button` hooked to present a [`FlutterViewController`][].
The `FlutterViewController` uses the `FlutterEngine` instance

> passed in as an `EnvironmentObject`.

Reviewer 2: Nit: This sounds like `FlutterViewController` requires an environment object. 
Also it's the FlutterDependencies, not flutter engine that gets injected. 

Maybe split into 2 sentences to make it more clear: 
1. the flutter view controller's constructor takes the pre-warmed engine as an argument 
2. this particular example uses environment object to inject flutter dependencies, which contains the flutter engine (this sentence is not as important, since you can also use singleton instead of dependency injection).

Author: Good call, thanks!

<!--code-excerpt "ContentView.swift" title-->
```swift
import SwiftUI
import Flutter

struct ContentView: View {

>   // Flutter engine is passed in as an EnvironmentObject.

Reviewer 2: similar here, technically it's the flutter dependencies, not flutter engine.

Author: Thanks!

  @EnvironmentObject var flutterDependencies: FlutterDependencies

  // Button is created to call the showFlutter function when pressed.
  var body: some View {
    Button("Show Flutter!") {
      showFlutter()
    }
  }

func showFlutter() {
    // Get the RootViewController.
    guard
      let windowScene = UIApplication.shared.connectedScenes
        .first(where: { $0.activationState == .foregroundActive && $0 is UIWindowScene }) as? UIWindowScene,
      let window = windowScene.windows.first(where: \.isKeyWindow),
      let rootViewController = window.rootViewController
    else { return }

    // Create the FlutterViewController.
    let flutterViewController = FlutterViewController(

>       engine: flutterDependencies.npsFlutterEngine,

Reviewer 2: IIRC "nps" is from the news feed project, so readers of this doc may not have context.

### Show a FlutterViewController with your FlutterEngine

{% samplecode vc %}

{% sample SwiftUI %}
The following example shows a generic `ContentView` with a
`Button` hooked to present a [`FlutterViewController`][].
The `FlutterViewController` constructor takes the pre-warmed 
`FlutterEngine` as an argument. `FlutterEngine` is passed in 
as an `EnvironmentObject` via `flutterDependencies`.

<!--code-excerpt "ContentView.swift" title-->
```swift
import SwiftUI
import Flutter

struct ContentView: View {

>   // Flutter dependencies are passed in as an EnvironmentObject.

Reviewer 3: ```suggestion
  // Flutter dependencies are passed in an EnvironmentObject.
```

`FlutterViewController` without an engine reference.

{% samplecode no-engine-vc %}
{% sample SwiftUI %}
```swift
import SwiftUI
import Flutter

struct ContentView: View {

>   // No EnvironmentObject needed.

Reviewer 2: nit: remove? it may need environment object for other dependencies.

Author: Sounds good

  (which can only be detected in the AppDelegate) to
  Flutter for scroll-to-top behavior.

> ### Creating a FlutterAppDelegate subclass

Reviewer 2: it's very possible that they already have an app delegate. maybe also document what to do if you can't inherit from flutter app delegate? iirc they will have to copy some functionalities from flutter app delegate over.

Author: That seems to be covered in the next section, but the example is Objective-C. I'm not really sure what the Swift equivalent looks like. Could you provide some guidance?

Reviewer 2: > That seems to be covered in the next section, but the example is Objective-C.

can you point me there? i can help to translate it into swift.

Author: Yes, the current section is here: https://docs.flutter.dev/development/add-to-app/ios/add-flutter-screen?tab=engine-objective-c-tab#using-the-flutterappdelegate

In the new version of the doc it's under a subheader ### If you can't directly make FlutterAppDelegate as subclass

Reviewer 2: I translated this section to Swift here: 
https://github.com/hellohuanlin/put-flutter-to-work/commit/7515c37309cc839d24d969bf2f677bb897c49311

Also note that the existing objc section about the status bar seems outdated: https://github.com/flutter/flutter/issues/63051

So i did not translate that.

Reviewer 2: @cyanglaz may know what happens to this status bar API.

Author: I don't think I have access to your version of put-flutter-to-work, can you grant me access?

Reviewer 2: Done

Reviewer 4: It was removed in https://github.com/flutter/engine/pull/16820, not sure about the details, but it looks like people don't need to use this API anymore. @gaaclarke to confirm.

Author: Cool thanks, I'm going to take it out for now.

Creating a subclass of the the `FlutterAppDelegate` in UIKit apps was shown 
in the [Start a FlutterEngine and FlutterViewController section][]. 
In a SwiftUI app, you can create a subclass of the 
`FlutterAppDelegate` that conforms to the `ObservabeObject` protocol as follows:

```swift
import SwiftUI
import Flutter
import FlutterPluginRegistrant

class AppDelegate: FlutterAppDelegate, ObservableObject {
  let flutterEngine = FlutterEngine(name: "my flutter engine")

  override func application(
    _ application: UIApplication,
    didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {

>       // Runs the default Dart entrypoint with a default Flutter route.

Reviewer 2: not sure if i understand this comment. seems like a in-depth topic that may not be suitable for an integration guide.

Author: This comment was used in the existing Swift example so I was just trying to be consistent. The section near the bottom of this doc talks through the entry points so I think it's fine to leave

Reviewer 2: gotcha. my flutter knowledge is limited, but i guess it's fine if it's already there.

      flutterEngine.run();
      // Used to connect plugins (only if you have plugins with iOS platform code).

>       GeneratedPluginRegistrant.register(with: self.flutterEngine);

Reviewer 2: should we cover plugins? do we expect developers at this stage to know what plugins are?

Author: I can maybe link to the plugin docs

      return true;
    }
}

@main
struct MyApp: App {
//  Use this property wrapper to tell SwiftUI

> //  it should use the AppDelegate class for the application delegate

Reviewer 2: nit: maybe remove? we don't have to teach swiftui here

Author: I just thought it might be useful in case people are not using app delegate with SwiftUI, since it isn't required

Reviewer 2: gotcha

  @UIApplicationDelegateAdaptor(AppDelegate.self) var appDelegate

  var body: some Scene {
      WindowGroup {
        ContentView()
      }
  }
}
```

Then, in your view, the `AppDelegate`is accessible as an `EnvionrmentObject`.

```swift
import SwiftUI
import Flutter

struct ContentView: View {
  // Access the AppDelegate using an EnvironmentObject.
  @EnvironmentObject var appDelegate: AppDelegate

  var body: some View {
    Button("Show Flutter!") {
      openFlutterApp()
    }
  }

func openFlutterApp() {
    let flutterEngine = appDelegate.flutterEngine

    // Get the RootViewController.
    guard
      let windowScene = UIApplication.shared.connectedScenes
        .first(where: { $0.activationState == .foregroundActive && $0 is UIWindowScene }) as? UIWindowScene,
      let window = windowScene.windows.first(where: \.isKeyWindow),
      let rootViewController = window.rootViewController
    else { return }

    // Create the FlutterViewController.
    let flutterViewController = FlutterViewController(

>       engine: flutterEngine,

Reviewer 2: nit: can just be `engine: appDelegate.flutterEngine`

Author: Sounds good
